src/formula/variant.cpp

```diff
@@ -212,194 +188,123 @@ variant_iterator& variant_iterator::operator=(const variant_iterator& that)
+	} else if(type_== TYPE_NULL &&  that.type_ == TYPE_NULL) {
 		return true;
-	else
-		return false;
+	}
+
+	return false;
 }
 
 bool variant_iterator::operator!=(const variant_iterator& that) const
 {
-	if (type_ == TYPE_LIST)
-	{
-		if (that.type_ != TYPE_LIST)
-			return true;
-		return list_iterator_ != that.list_iterator_;
-	} else if (type_ == TYPE_MAP)
-	{
-		if (that.type_ != TYPE_MAP)
-			return true;
-		return map_iterator_ != that.map_iterator_;
-	} else if (type_ == TYPE_NULL &&  that.type_ == TYPE_NULL )
-		return false;
-	else
-		return true;
+	return !operator==(that);
 }
 
-void variant::release()
-{
-	switch(type_) {
-	case TYPE_LIST:
-		delete list_;
-		break;
-	case TYPE_STRING:
-		delete string_;
-		break;
-	case TYPE_MAP:
-		delete map_;
-		break;
-	// These are not used here, add them to silence a compiler warning.
-	case TYPE_CALLABLE:
-	case TYPE_NULL:
-	case TYPE_DECIMAL:
-	case TYPE_INT :
-		break;
-	}
-}
 
-std::string variant::type_string() const {
-	return variant_type_to_string(type_);
-}
-
-variant::variant() : type_(TYPE_NULL), int_value_(0)
-{}
-
-variant::variant(int n) : type_(TYPE_INT), int_value_(n)
+variant::variant()
+	: value_()
 {}
 
-variant::variant(int n, variant::DECIMAL_VARIANT_TYPE /*type*/) : type_(TYPE_DECIMAL), decimal_value_(n)
-{}
-
-variant::variant(double n, variant::DECIMAL_VARIANT_TYPE /*type*/) : type_(TYPE_DECIMAL) {
-	n *= 1000;
-	decimal_value_ = static_cast<int>(n);
+variant::variant(int n)
+	: value_(std::make_shared<game_logic::variant_int>(n))
+{
+	assert(value_.get());
+}
 
-	n -= decimal_value_;
+variant::variant(int n, variant::DECIMAL_VARIANT_TYPE)
+	: value_(std::make_shared<game_logic::variant_decimal>(n))
+{
+	assert(value_.get());
+}
 
-	if(n > 0.5)
-		decimal_value_++;
-	else if(n < -0.5)
-		decimal_value_--;
+variant::variant(double n, variant::DECIMAL_VARIANT_TYPE)
+	: value_(std::make_shared<game_logic::variant_decimal>(n))
+{
+	assert(value_.get());
 }
 
 variant::variant(const game_logic::formula_callable* callable)
-	: type_(TYPE_CALLABLE), callable_(callable)
+	: value_(std::make_shared<game_logic::variant_callable>(callable))
 {
-	assert(callable_);
+	assert(value_.get());
 }
 
-variant::variant(std::vector<variant>* array)
-    : type_(TYPE_LIST)
+variant::variant(std::vector<variant>* vec)
+    : value_((std::make_shared<game_logic::variant_list>(*vec)))
 {
-	assert(array);
-	list_ = new std::vector<variant>(*array);
+	assert(value_.get());
 }
 
 variant::variant(const std::string& str)
-	: type_(TYPE_STRING)
+	: value_(std::make_shared<game_logic::variant_string>(str))
 {
-	string_ = new std::string(str);
+	assert(value_.get());
 }
 
 variant::variant(std::map<variant,variant>* map)
-    : type_(TYPE_MAP)
+	: value_((std::make_shared<game_logic::variant_map>(*map)))
 {
-	assert(map);
-	map_ = new std::map<variant,variant>(*map);
+	assert(value_.get());
 }
 
 variant::variant(const variant& v)
-    : type_(v.type_)
```

Comment:
Oh huh, looks like the old copy constructor had significant memory management issues.